drafts/generating-css-from-scratch-with-postcss.md

```diff
@@ -0,0 +1,113 @@
+
+<!-- excerpt -->
+
+### PostCSS
+
+[PostCSS](https://postcss.org/) is `A tool for transforming CSS with JavaScript`. It's widely use in the industry for things like [auto-prefixing CSS](https://autoprefixer.github.io/).
+
+As the tag line says, PostCSS main business is *transforming* CSS, not generating it from scratch as I wanted to do. Looking at the PostCSS codebase I noticed [a test](https://github.com/postcss/postcss/blob/master/test/postcss.test.js#L116) named `it allows to build own CSS` so I figured it should be possible!
+
+Fast forward a couple of hours diving through the PostCSS codebase, reading the API docs and several PostCSS plugins, I had a working solution.
+
+### Generating CSS from scratch
+
+To generate CSS with PostCSS you first need to build up an AST:
+
+```js
+const postcss = require("postcss");
+
+const fontFamily = "My Family";
+const root = postcss.root();
+
+const bodyRule = postcss.rule({ selector: "body" }).append(
+  postcss.decl({
+    prop: "font-family",
+    value: `"${fontFamily}"`
+  })
+);
+
+const fontFace = postcss.atRule({ name: "font-face" }).append([
+   postcss.decl({ prop: "font-family", value: `"${fontFamily}"` }),
+   postcss.decl({ prop: "src", value: "url(./fonts/myfont.woff2)" })
+]);
+
+root.append([
+  fontFace,
+  bodyRule
+]);
+```
+
+The CSS code can then be generated from the CSS with `toString()`, resulting in this:
+
+```css
+@font-face {
+    font-family: "My Family";
+    src: url(./fonts/myfont.woff2)
+}
+body {
+    font-family: "My Family"
+}
+```
+
+A couple of things to notice:
+
+* PostCSS does not provide much help in generating proper declaration values [out of the box](https://github.com/TrySound/postcss-value-parser). For example quoting a `font-family` value in case it contains multiple words is not handled by PostCSS automatically.
+* It's easy to conditionally generate properties, rules, declarations etc. The API follows a typical builder pattern, which makes it easy to conditionally call `append`, or not.
+* The CSS output looks opinionated. There are no semicolons after the last declaration and no newlines between rules.
+
+Fortunately, PostCSS is architected quite well and allows you to provide your own ["Stringifier"](http://api.postcss.org/global.html#stringifier). I didn't find much documentation or guidance on this though, but after a bit of code diving I settled on this:
+
+```js
+const Stringifier = require("postcss/lib/stringifier");
+
+class PrettyStringifier extends Stringifier {
+  static new() {
+    return (node, builder) => {
+      let str = new this(builder);
+      str.stringify(node);
+    };
+  }
+
+  constructor(builder) {
+    super(builder);
+  }
+
+  rule(node) {
+    if (node.prev()) {
+      // Add a newline after a rule, if it's preceded by another rule.
+      this.builder("\n", node);
+    }
+
+    return super.rule(node);
+  }
+
+  decl(node) {
+    return super.decl(node, true /* force semicolon */);
```

Comment:
Nitpick: mixing `//` and `/* */` comment, I get why it's happening here but it looks kinda willy-nilly on first sight.